From 37e2b306d5b21d974f3fcb0e66d2709b7877cbaa Mon Sep 17 00:00:00 2001
From: Stefan Berger <stefanb@linux.ibm.com>
Date: Tue, 3 Jan 2023 18:05:51 -0500
Subject: [PATCH 83/87] security: Pass security.* xattrs on to EVM if
 CAP_MAC_ADMIN is set

Allow the handling of all security.* xattrs in user namespaces by
EVM if CAP_MAC_ADMIN is set and the inode is fully owned by the
namespace. security.* xattr can now be altered in a user namespace
if a file is fully mapped (uid and gid) in the user namespace and
if EVM is compileed in and the user has CAP_MAC_ADMIN in the user
namespace.

Signed-off-by: Stefan Berger <stefanb@linux.ibm.com>
---
 include/linux/security.h   |  5 +++--
 security/commoncap.c       | 29 +++++++++++++++++++++++++++--
 security/security.c        |  3 ++-
 security/selinux/hooks.c   |  3 ++-
 security/smack/smack_lsm.c |  3 ++-
 5 files changed, 36 insertions(+), 7 deletions(-)

diff --git a/include/linux/security.h b/include/linux/security.h
index 5b67f208f7de..b3cb2f3bca74 100644
--- a/include/linux/security.h
+++ b/include/linux/security.h
@@ -151,7 +151,8 @@ extern int cap_capset(struct cred *new, const struct cred *old,
 		      const kernel_cap_t *inheritable,
 		      const kernel_cap_t *permitted);
 extern int cap_bprm_creds_from_file(struct linux_binprm *bprm, struct file *file);
-int cap_inode_setxattr(struct dentry *dentry, const char *name,
+int cap_inode_setxattr(struct user_namespace *mnt_userns,
+		       struct dentry *dentry, const char *name,
 		       const void *value, size_t size, int flags);
 int cap_inode_removexattr(struct user_namespace *mnt_userns,
 			  struct dentry *dentry, const char *name);
@@ -878,7 +879,7 @@ static inline int security_inode_setxattr(struct user_namespace *mnt_userns,
 		struct dentry *dentry, const char *name, const void *value,
 		size_t size, int flags)
 {
-	return cap_inode_setxattr(dentry, name, value, size, flags);
+	return cap_inode_setxattr(mnt_userns, dentry, name, value, size, flags);
 }
 
 static inline int security_inode_set_acl(struct user_namespace *mnt_userns,
diff --git a/security/commoncap.c b/security/commoncap.c
index 5f1f45702ea5..286736956b4e 100644
--- a/security/commoncap.c
+++ b/security/commoncap.c
@@ -979,6 +979,7 @@ int cap_bprm_creds_from_file(struct linux_binprm *bprm, struct file *file)
 
 /**
  * cap_inode_setxattr - Determine whether an xattr may be altered
+ * @mnt_userns:	User namespace of the mount the inode was found from
  * @dentry: The inode/dentry being altered
  * @name: The name of the xattr to be changed
  * @value: The value that the xattr will be changed to
@@ -991,7 +992,8 @@ int cap_bprm_creds_from_file(struct linux_binprm *bprm, struct file *file)
  * This is used to make sure security xattrs don't get updated or set by those
  * who aren't privileged to do so.
  */
-int cap_inode_setxattr(struct dentry *dentry, const char *name,
+int cap_inode_setxattr(struct user_namespace *mnt_userns,
+		       struct dentry *dentry, const char *name,
 		       const void *value, size_t size, int flags)
 {
 	struct user_namespace *user_ns = dentry->d_sb->s_user_ns;
@@ -1014,6 +1016,17 @@ int cap_inode_setxattr(struct dentry *dentry, const char *name,
 	if (strcmp(name, XATTR_NAME_IMA) == 0)
 		return 0;
 
+	/*
+	 * All 'security.' xattrs in a user namespace will be handled by
+	 * EVM with CAP_MAC_ADMIN.
+	 */
+	if (current_user_ns() != &init_user_ns
+	    && IS_ENABLED(CONFIG_EVM)
+	    && mac_admin_ns_capable(current_user_ns())
+	    && privileged_wrt_inode_uidgid(current_user_ns(), mnt_userns,
+					   d_backing_inode(dentry)))
+		return 0;
+
 	if (!ns_capable(user_ns, CAP_SYS_ADMIN))
 		return -EPERM;
 	return 0;
@@ -1042,15 +1055,16 @@ int cap_inode_removexattr(struct user_namespace *mnt_userns,
 			  struct dentry *dentry, const char *name)
 {
 	struct user_namespace *user_ns = dentry->d_sb->s_user_ns;
+	struct inode *inode;
 
 	/* Ignore non-security xattrs */
 	if (strncmp(name, XATTR_SECURITY_PREFIX,
 			XATTR_SECURITY_PREFIX_LEN) != 0)
 		return 0;
 
+	inode = d_backing_inode(dentry);
 	if (strcmp(name, XATTR_NAME_CAPS) == 0) {
 		/* security.capability gets namespaced */
-		struct inode *inode = d_backing_inode(dentry);
 		if (!inode)
 			return -EINVAL;
 		if (!capable_wrt_inode_uidgid(mnt_userns, inode, CAP_SETFCAP))
@@ -1064,6 +1078,17 @@ int cap_inode_removexattr(struct user_namespace *mnt_userns,
 	if (strcmp(name, XATTR_NAME_IMA) == 0)
 		return 0;
 
+	/*
+	 * All 'security.' xattrs in a user namespace will be handled by
+	 * EVM with CAP_MAC_ADMIN.
+	 */
+	if (current_user_ns() != &init_user_ns
+	    && IS_ENABLED(CONFIG_EVM)
+	    && mac_admin_ns_capable(current_user_ns())
+	    && privileged_wrt_inode_uidgid(current_user_ns(), mnt_userns,
+					   inode))
+		return 0;
+
 	if (!ns_capable(user_ns, CAP_SYS_ADMIN))
 		return -EPERM;
 	return 0;
diff --git a/security/security.c b/security/security.c
index 2b1f9ec8dea9..44010fb51ddc 100644
--- a/security/security.c
+++ b/security/security.c
@@ -1391,7 +1391,8 @@ int security_inode_setxattr(struct user_namespace *mnt_userns,
 			    size, flags);
 
 	if (ret == 1)
-		ret = cap_inode_setxattr(dentry, name, value, size, flags);
+		ret = cap_inode_setxattr(mnt_userns, dentry, name, value, size,
+					 flags);
 	if (ret)
 		return ret;
 	ret = ima_inode_setxattr(dentry, name, value, size);
diff --git a/security/selinux/hooks.c b/security/selinux/hooks.c
index 3c5be76a9199..2e8651bce615 100644
--- a/security/selinux/hooks.c
+++ b/security/selinux/hooks.c
@@ -3157,7 +3157,8 @@ static int selinux_inode_setxattr(struct user_namespace *mnt_userns,
 	int rc = 0;
 
 	if (strcmp(name, XATTR_NAME_SELINUX)) {
-		rc = cap_inode_setxattr(dentry, name, value, size, flags);
+		rc = cap_inode_setxattr(mnt_userns, dentry, name, value, size,
+				        flags);
 		if (rc)
 			return rc;
 
diff --git a/security/smack/smack_lsm.c b/security/smack/smack_lsm.c
index 9a82a15685d1..5b0a6584839c 100644
--- a/security/smack/smack_lsm.c
+++ b/security/smack/smack_lsm.c
@@ -1248,7 +1248,8 @@ static int smack_inode_setxattr(struct user_namespace *mnt_userns,
 		    strncmp(value, TRANS_TRUE, TRANS_TRUE_SIZE) != 0)
 			rc = -EINVAL;
 	} else
-		rc = cap_inode_setxattr(dentry, name, value, size, flags);
+		rc = cap_inode_setxattr(mnt_userns, dentry, name, value, size,
+					flags);
 
 	if (check_priv && !smack_privileged(CAP_MAC_ADMIN))
 		rc = -EPERM;
-- 
2.37.3

